File: 2021/Day_01/aoc_2021_day01.m
soundings = parse_soundings('./puzzle_input.txt');
% soundings = [199, 200, 208, 210, 200, 207, 240, 269, 260, 263];

fprintf('Part One: %u\n', count_adjacent(soundings));
fprintf('Part Two: %u\n', count_sums(soundings, 3));


function [soundings] = parse_soundings(filepath)
% Parse the provided depth soundings into an n x 1 vector.
arguments
    filepath char {mustBeFile}
end

fid = fopen(filepath, 'r');
soundings = fscanf(fid, '%u');
fclose(fid);
end

function [n_ascending] = count_adjacent(soundings)
% Count the number of times the depth sounding increases from the previous measurement.
arguments
    soundings (:, 1) {mustBeNumeric, mustBeNonempty, mustBeVector}
end

n_ascending = sum(diff(soundings) > 0);
end

function [n_ascending] = count_sums(soundings, width)
% Count the number of times the sum of a sliding window increases from the previous window's sum.
arguments
    soundings (:, 1) {mustBeNumeric, mustBeNonempty, mustBeVector}
    width (1, 1) {mustBeInteger} = 3
end

n_windows = length(soundings) - width + 1;
windows = zeros(n_windows, 1);
for ii = 1:n_windows
    windows(ii) = sum(soundings(ii:(ii+width-1)));
end

n_ascending = sum(diff(windows) > 0);
end

File: 2021/Day_01/aoc_2021_day01.py
from pathlib import Path

import more_itertools as mi


def count_ascending(soundings: list[int], width: int = 1) -> int:
    """
    Count the number of times the sum of a sliding window increases from the previous window.

    A window width of 1 can be used to consider adjacent values.
    """
    n_ascending = 0
    # When comparing the sums of adjacent sliding windows, they share the middle value(s) so the
    # only delta we need to consider is between the valuies on the boundary of their combined window
    for left, *_, right in mi.sliding_window(soundings, width + 1):
        if right > left:
            n_ascending += 1

    return n_ascending


if __name__ == "__main__":
    puzzle_input_file = Path("./puzzle_input.txt")
    puzzle_input = [int(line) for line in puzzle_input_file.read_text().splitlines()]

    n_ascending_1 = count_ascending(puzzle_input, 1)
    print(f"Part One: {n_ascending_1}")

    n_ascending_2 = count_ascending(puzzle_input, 3)
    print(f"Part Two: {n_ascending_2}")
